File: sdk/js/strip.js
/* The privacy strip — what leaves the machine when a record publishes.
 *
 * A port of string/app/strip.py, the canonical copy the String's publisher,
 * scripts/promote.py and scripts/export_public.py all use, so a client that
 * publishes directly (Pocket, Easel, Loom) and the String cannot disagree
 * about what is private. tests/fixtures/strip-cases.json is run by both
 * languages so they keep agreeing.
 *
 * Allow lists all the way down: a field not named here does not publish, at
 * any depth, so a new lexicon field stays private until someone decides
 * otherwise and adds a fixture saying so. Geo, provenance, device ids,
 * mintIds, local media refs and resolver bookkeeping never leave. A person
 * ref — or a ref of any type other than work, event, venue or concept —
 * publishes only with a DID or an external identifier: a bare name may be a
 * private individual. A `did` or `creatorDid` that does not match the DID
 * pattern counts as absent and does not publish.
 *
 * CANONICAL COPY. Apps carry copies; copy outward from here.
 */

const ANNOTATION = 'com.cultureblocs.annotation';
const ROLES = ['subject', 'mention'];
const NON_PERSON_TYPES = ['work', 'event', 'venue', 'concept'];
const BEAD_KEEP = ['createdAt', 'kind', 'note'];
const STRAND_KEEP = ['createdAt', 'title', 'narrative', 'day'];

/* Local-only machinery on an otherwise public-by-intent record. `provenance`
 * is device and app internals; `media` points at files on the author's own
 * String, which no stranger can resolve. */
const LOCAL_ONLY = ['provenance', 'media'];

/* ATProto DID syntax — the same pattern as sdk/js/lexicon.js and
 * string/app/lexicon.py DID_RE. Kept here, not imported, because apps copy
 * this file on its own. */
const DID_RE = /^did:[a-z0-9]+:[a-zA-Z0-9._:%-]+$/;

const isObject = (v) => typeof v === 'object' && v !== null && !Array.isArray(v);
const str = (v) => typeof v === 'string' && v !== '';
const list = (v) => (Array.isArray(v) ? v : []);
const did = (v) => typeof v === 'string' && DID_RE.test(v);

function pick(d, keys) {
  const out = {};
  for (const k of keys) if (str(d[k])) out[k] = d[k];
  return out;
}

/* Drop a creatorDid that pick kept but that is not a DID. */
function dropBadCreatorDid(out) {
  if ('creatorDid' in out && !did(out.creatorDid)) delete out.creatorDid;
  return out;
}

function requireType(body) {
  if (!body || typeof body['$type'] !== 'string') {
    throw new Error('record body has no $type');   // Python raises KeyError here
  }
  return body['$type'];
}

/* linkRefs: uri and title only. */
export const stripLinks = (links) =>
  list(links).filter((l) => isObject(l) && str(l.uri)).map((l) => pick(l, ['uri', 'title']));

export const stripTags = (tags) => list(tags).filter(str);

/* externalIds: scheme, id and uri; entries without scheme and id are dropped. */
export const stripExternalIds = (ids) =>
  list(ids).filter((e) => isObject(e) && str(e.scheme) && str(e.id))
    .map((e) => pick(e, ['scheme', 'id', 'uri']));

/* The public form of one #ref, or null if it must not publish. `anchored` is
 * false for presentation refs, which have no text to anchor into. */
export function stripRef(ref, anchored = true) {
  if (!isObject(ref) || !str(ref.type)) return null;
  const descriptor = ref.descriptor;
  if (!isObject(descriptor) || !str(descriptor.label)) return null;
  const out = {
    type: ref.type,
    role: ROLES.includes(ref.role) ? ref.role : 'mention',
    descriptor: dropBadCreatorDid(pick(descriptor, ['label', 'creator', 'creatorDid', 'date'])),
  };
  if (did(ref.did)) out.did = ref.did;
  const ids = stripExternalIds(ref.externalIds);
  if (ids.length) out.externalIds = ids;
  if (!NON_PERSON_TYPES.includes(ref.type) && !('did' in out) && !ids.length) return null; // unknown types fail closed
  const index = ref.index;
  if (anchored && isObject(index) && Number.isInteger(index.byteStart) && Number.isInteger(index.byteEnd)) {
    out.index = { byteStart: index.byteStart, byteEnd: index.byteEnd };
  }
  return out;
}

export const stripRefs = (refs) => list(refs).map((r) => stripRef(r)).filter((r) => r !== null);

export function stripPresentation(presentation) {
  if (!isObject(presentation)) return null;
  const out = pick(presentation, ['format']);
  for (const key of ['venueRef', 'eventRef']) {
    const ref = stripRef(presentation[key], false);
    if (ref !== null) out[key] = ref;
  }
  return Object.keys(out).length ? out : null;
}

/* Deprecated #workRef on annotations: identifiers and descriptors, never `image`. */
export function stripWorkRef(work) {
  if (!isObject(work)) return null;
  const out = dropBadCreatorDid(
    pick(work, ['title', 'creator', 'date', 'wikidata', 'linkedArt', 'creatorDid']));
  const acc = work.accession;
  if (isObject(acc) && str(acc.institution) && str(acc.id)) out.accession = pick(acc, ['institution', 'id']);
  return out;
}

function common(body, out) {
  const tags = stripTags(body.tags);
  if (tags.length) out.tags = tags;
  const links = stripLinks(body.links);
  if (links.length) out.links = links;
  const refs = stripRefs(body.refs);
  if (refs.length) out.refs = refs;
  return out;
}

/* A bead or annotation, as it publishes inside a strand.
 *
 * `subject` is reduced to its name and nothing else: the whole point is that
 * "Tate Modern" publishes while the coordinates that would place you in it
 * do not. A subject with no name drops entirely rather than publishing an
 * empty husk. `images` are imageRefs the caller has already uploaded. */
export function stripBead(body, { images = null } = {}) {
  const out = { $type: requireType(body) };
  for (const k of BEAD_KEEP) if (str(body[k])) out[k] = body[k];
  if (isObject(body.subject) && str(body.subject.name)) out.subject = { name: body.subject.name };
  if (body.$type === ANNOTATION) {
    const work = stripWorkRef(body.work);
    if (work !== null) out.work = work;
  }
  const presentation = stripPresentation(body.presentation);
  if (presentation !== null) out.presentation = presentation;
  common(body, out);
  if (images && images.length) out.images = images;
  return out;
}

/* A strand. `items` are the at:// refs of the beads already published above
 * it — never the local spine:// uris, which is why they are passed in rather
 * than copied from the body. A target that bundles items inline (the static
 * export) passes null, and the field is omitted. */
export function stripStrand(body, items) {
  const out = { $type: requireType(body) };
  for (const k of STRAND_KEEP) if (str(body[k])) out[k] = body[k];
  if (isObject(body.place) && str(body.place.name)) out.place = { name: body.place.name };
  common(body, out);
  if (items !== null && items !== undefined) out.items = items;
  return out;
}

/* Records that are public by intent — creative claims, venue listings,
 * calendar events. These are written to be read by strangers, so the body
 * publishes as authored, minus the local-only machinery.
 *
 * Note the asymmetry with stripBead: this is a DENY list, so a field added
 * to one of those lexicons publishes automatically. That is deliberate for
 * records whose purpose is to be read, and wrong for a diary bead — hence
 * the allow lists above. A venue's address and coordinates are the point of
 * the record and stay. */
export function stripPublic(body) {
  const out = {};
  for (const [k, v] of Object.entries(body)) {
    if (!LOCAL_ONLY.includes(k)) out[k] = v;
  }
  return out;
}

/* The hash the String stores as publishedHash, and the shape Easel calls
 * canonicalJSON. Sorted keys, no whitespace — byte-identical to Python's
 * json.dumps(obj, sort_keys=True, separators=(",", ":")). */
export function canonicalJSON(obj) {
  return JSON.stringify(sortKeys(obj));
}

function sortKeys(v) {
  if (Array.isArray(v)) return v.map(sortKeys);
  if (v && typeof v === 'object') {
    const out = {};
    for (const k of Object.keys(v).sort()) out[k] = sortKeys(v[k]);
    return out;
  }
  return v;
}

/* sha256 hex of the canonical form. Matches publisher.content_hash.
 * Async because WebCrypto is; Node's webcrypto satisfies the same call. */
export async function contentHash(obj) {
  const bytes = new TextEncoder().encode(canonicalJSON(obj));
  const digest = await crypto.subtle.digest('SHA-256', bytes);
  return [...new Uint8Array(digest)].map(b => b.toString(16).padStart(2, '0')).join('');
}

File: string/app/strip.py
"""The publish strip: the one canonical copy.

What may leave the String when a record is published. Allowlists all the
way down — a field not named here does not publish, at any depth, so a new
lexicon field stays private until someone decides otherwise and adds a
fixture saying so. Geo, provenance, local media and resolver bookkeeping
never leave.

Used by publisher.py, scripts/promote.py and scripts/export_public.py.
Ported to sdk/js/strip.js; both run tests/fixtures/strip-cases.json.
"""
from __future__ import annotations

import hashlib
import json

from .lexicon import DID_RE

ANNOTATION = "com.cultureblocs.annotation"
ROLES = ("subject", "mention")
NON_PERSON_TYPES = ("work", "event", "venue", "concept")
BEAD_FIELDS = ("createdAt", "kind", "note")
STRAND_FIELDS = ("createdAt", "title", "narrative", "day")


def _str(v) -> bool:
    return isinstance(v, str) and v != ""


def _int(v) -> bool:
    return isinstance(v, int) and not isinstance(v, bool)


def _did(v) -> bool:
    """A DID in ATProto syntax. Anything else in a DID field is treated as
    absent: a made-up `did` must not vouch for a bare name."""
    return isinstance(v, str) and DID_RE.fullmatch(v) is not None


def _pick(d: dict, keys: tuple[str, ...]) -> dict:
    return {k: d[k] for k in keys if _str(d.get(k))}


def _drop_bad_creator_did(out: dict) -> dict:
    """Drop a creatorDid that `_pick` kept but that is not a DID."""
    if "creatorDid" in out and not _did(out["creatorDid"]):
        del out["creatorDid"]
    return out


def _head(body: dict) -> dict:
    return {"$type": body["$type"]}   # KeyError, like sdk/js/strip.js, rather than an untyped record


def strip_links(links) -> list[dict]:
    """linkRefs: uri and title only."""
    return [_pick(link, ("uri", "title")) for link in (links if isinstance(links, list) else [])
            if isinstance(link, dict) and _str(link.get("uri"))]


def strip_tags(tags) -> list[str]:
    return [t for t in (tags if isinstance(tags, list) else []) if _str(t)]


def strip_external_ids(ids) -> list[dict]:
    """externalIds: scheme, id and uri; entries without scheme and id are dropped."""
    return [_pick(e, ("scheme", "id", "uri")) for e in (ids if isinstance(ids, list) else [])
            if isinstance(e, dict) and _str(e.get("scheme")) and _str(e.get("id"))]


def strip_ref(ref, anchored: bool = True) -> dict | None:
    """The public form of one #ref, or None if it must not publish.

    Args:
        ref: A ref as stored locally.
        anchored: False for refs that have no text to anchor into
            (presentation.venueRef / eventRef); their index is dropped.

    Returns:
        The stripped ref, or None when it lacks a type or label, or when it
        is a person ref — or a ref of any type other than work, event,
        venue or concept — with neither a DID nor an external identifier.
        A `did` or `descriptor.creatorDid` that does not match the DID
        pattern is dropped, as if absent.
    """
    if not isinstance(ref, dict) or not _str(ref.get("type")):
        return None
    descriptor = ref.get("descriptor")
    if not isinstance(descriptor, dict) or not _str(descriptor.get("label")):
        return None
    out = {
        "type": ref["type"],
        "role": ref["role"] if ref.get("role") in ROLES else "mention",
        "descriptor": _drop_bad_creator_did(
            _pick(descriptor, ("label", "creator", "creatorDid", "date"))),
    }
    if _did(ref.get("did")):
        out["did"] = ref["did"]
    ids = strip_external_ids(ref.get("externalIds"))
    if ids:
        out["externalIds"] = ids
    if ref["type"] not in NON_PERSON_TYPES and "did" not in out and not ids:
        return None  # a bare name may be a private individual; unknown types fail closed
    index = ref.get("index")
    if anchored and isinstance(index, dict) \
            and _int(index.get("byteStart")) and _int(index.get("byteEnd")):
        out["index"] = {"byteStart": index["byteStart"], "byteEnd": index["byteEnd"]}
    return out


def strip_refs(refs) -> list[dict]:
    kept = (strip_ref(r) for r in (refs if isinstance(refs, list) else []))
    return [r for r in kept if r is not None]


def strip_presentation(presentation) -> dict | None:
    if not isinstance(presentation, dict):
        return None
    out = _pick(presentation, ("format",))
    for key in ("venueRef", "eventRef"):
        ref = strip_ref(presentation.get(key), anchored=False)
        if ref is not None:
            out[key] = ref
    return out or None


def strip_work_ref(work) -> dict | None:
    """Deprecated #workRef on annotations: identifiers and descriptors, never `image`."""
    if not isinstance(work, dict):
        return None
    out = _drop_bad_creator_did(
        _pick(work, ("title", "creator", "date", "wikidata", "linkedArt", "creatorDid")))
    accession = work.get("accession")
    if isinstance(accession, dict) and _str(accession.get("institution")) \
            and _str(accession.get("id")):
        out["accession"] = _pick(accession, ("institution", "id"))
    return out


def _common(body: dict, out: dict) -> dict:
    tags = strip_tags(body.get("tags"))
    if tags:
        out["tags"] = tags
    links = strip_links(body.get("links"))
    if links:
        out["links"] = links
    refs = strip_refs(body.get("refs"))
    if refs:
        out["refs"] = refs
    return out


def strip_bead(body: dict, images: list | None = None) -> dict:
    """Public subset of a bead or annotation.

    Local `media` refs never publish; if the caller has uploaded them, they
    arrive as `images` — imageRefs carrying the blob plus alt text and
    dimensions — and are attached as given.
    """
    out = _head(body)
    for k in BEAD_FIELDS:
        if _str(body.get(k)):
            out[k] = body[k]
    subject = body.get("subject")
    if isinstance(subject, dict) and _str(subject.get("name")):
        out["subject"] = {"name": subject["name"]}
    if body.get("$type") == ANNOTATION:
        work = strip_work_ref(body.get("work"))
        if work is not None:
            out["work"] = work
    presentation = strip_presentation(body.get("presentation"))
    if presentation is not None:
        out["presentation"] = presentation
    _common(body, out)
    if images:
        out["images"] = images
    return out


def strip_strand(body: dict, items: list[dict] | None = None) -> dict:
    """Public subset of a strand. `items` are the published strongRefs,
    filled in at publish time; omitted for targets that bundle items inline."""
    out = _head(body)
    for k in STRAND_FIELDS:
        if _str(body.get(k)):
            out[k] = body[k]
    place = body.get("place")
    if isinstance(place, dict) and _str(place.get("name")):
        out["place"] = {"name": place["name"]}
    _common(body, out)
    if items is not None:
        out["items"] = items
    return out


def content_hash(obj: dict) -> str:
    """sha256 of the canonical JSON form (sorted keys, no whitespace); matches
    sdk/js/strip.js contentHash."""
    return hashlib.sha256(
        json.dumps(obj, sort_keys=True, separators=(",", ":")).encode()).hexdigest()


def drift_hash(body: dict) -> str:
    """The publishedHash of a bead or annotation: what `status` compares.

    Hashes the local body that was published, not the record the PDS got.
    That record's `images` hold blob refs that exist only after an upload, so
    they can never be recomputed from the String — hashing them made every
    bead with a photo look edited forever. Instead each local photo counts by
    its content-addressed file name, alt text and dimensions: adding, removing,
    replacing or re-describing a photo is drift; which host served it is not.
    A bead with no photos hashes exactly as content_hash(strip_bead(body)).
    """
    out = strip_bead(body)
    media = [{"file": m["uri"].rsplit("/", 1)[-1],
              **{k: m[k] for k in ("alt", "aspectRatio") if k in m}}
             for m in (body.get("media") if isinstance(body.get("media"), list) else [])
             if isinstance(m, dict) and _str(m.get("uri"))]
    if media:
        out["media"] = media
    return content_hash(out)


def strip_public(body: dict) -> dict:
    """Records that are public by intent (creative claims, venue listings).

    These are written to be read by strangers, so the body publishes as
    authored — minus local-only machinery: provenance (device/app internals)
    and `media` refs that point at files on the author's own String. A venue's
    address and coordinates are the point of the record and stay.
    """
    return {k: v for k, v in body.items() if k not in ("provenance", "media")}
